src/spawn.c

```diff
@@ -0,0 +1,1121 @@
+
+	if (spawned)
+	{
+		int *fd[3] = { stdin_fd, stdout_fd, stderr_fd };
+		int i;
+
+		for (i = 0; i < 3; i++)
+			if (fd[i])
+				fcntl(*fd[i], F_SETFL, O_NONBLOCK);
+	}
+
+	if (full_argv != argv)
+	{
+		full_argv[cl_argc] = NULL;
+		g_strfreev(full_argv);
+	}
+
+	return spawned;
+#endif  /* G_OS_WIN32 */
+}
+
+
+enum { OUT, ERR, IN };
+
+typedef struct _SpawnSource
+{
+	GSource source;
+	GPollFD std[3];
+	gchar *stdin_text;
+	const char *stdin_pos;
+	size_t stdin_remaining;
+#ifdef G_OS_WIN32
+	DWORD last_send_ticks;
+#endif
+	GIOChannel *ioc[2];
+	void (*io_cb[2])(const gchar *text, gpointer user_data);
+	gpointer io_data[2];
+	GPid pid;
+	void (*exit_cb)(gint status, gpointer user_data);
+	gpointer exit_data;
+	/* interactive */
+	GtkWidget *dialog;
+	const gchar *command;
+} SpawnSource;
+
+
+static void spawn_source_exit_cb(GPid pid, gint status, gpointer source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	if (sp->exit_cb)
+		sp->exit_cb(status, sp->exit_data);
+
+	g_spawn_close_pid(pid);
+	g_source_unref((GSource *) source);
+}
+
+
+static gboolean spawn_source_prepare(G_GNUC_UNUSED GSource *source, gint *timeout)
+{
+	*timeout = -1;
+	return FALSE;
+}
+
+
+#ifdef G_OS_WIN32
+static gboolean spawn_source_peek_pipe(GPollFD *poll)
+{
+	DWORD available;
+
+	return poll->fd != -1 && (!PeekNamedPipe((HANDLE) _get_osfhandle(poll->fd), NULL, 0, NULL,
+		&available, NULL) || available);
+}
+
+
+static gboolean spawn_source_check(GSource *source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	return (sp->std[IN].fd != -1 && GetTickCount() - sp->last_send_ticks >= 50) ||
+		spawn_source_peek_pipe(&sp->std[OUT]) || spawn_source_peek_pipe(&sp->std[ERR]);
+}
+#else  /* G_OS_WIN32 */
+static gboolean spawn_source_check(GSource *source)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+
+	return (sp->std[IN].fd != -1 && sp->std[IN].revents) ||
+		(sp->std[OUT].fd != -1 && sp->std[OUT].revents) ||
+		(sp->std[ERR].fd != -1 && sp->std[ERR].revents);
+}
+#endif  /* G_OS_WIN32 */
+
+
+static gboolean spawn_source_dispatch(GSource *source, G_GNUC_UNUSED GSourceFunc callback,
+	G_GNUC_UNUSED gpointer user_data)
+{
+	SpawnSource *sp = (SpawnSource *) source;
+	gchar *line;
+	int i;
+#ifdef G_OS_WIN32
+	DWORD status = STILL_ACTIVE;
+#else
+	int status = -1;
+#endif
+
+	if (sp->std[IN].fd != -1)
+	{
+		ssize_t count = write(sp->std[IN].fd, sp->stdin_pos, sp->stdin_remaining);
+
+		if (count > 0)
+		{
+			sp->stdin_remaining -= count;
+
+			if (sp->stdin_remaining)
+			{
+				sp->stdin_pos += count;
+			#ifdef G_OS_WIN32
+				sp->last_send_ticks = GetTickCount();
+			#endif
+			}
+			else
+			{
+				close(sp->std[IN].fd);
+				sp->std[IN].fd = -1;
+				g_free(sp->stdin_text);
+				sp->stdin_text = NULL;
+
+				if (sp->std[OUT].fd == -1 && sp->std[ERR].fd == -1)
+				{
+					g_source_ref(source);
+					g_child_watch_add(sp->pid, spawn_source_exit_cb, sp);
+					return FALSE;
+				}
+
+			#ifdef G_OS_UNIX
+				g_source_remove_poll(source, &sp->std[IN]);

```

Comment:
Which part exactly? Killing the text to send before sending it completely is not a good idea. As of the rest, it switches from poll mode to child watch mode if child's stdin was the only fd.